doc/hacking.rst

```diff
@@ -315,3 +315,161 @@ For automatically updated builds, you can install sphinx-autobuild, and run
 
 Nix users may drop into a development shell with the necessary dependencies for
 building docs ``nix develop .#doc``.
+
+General Guidelines
+==================
+
+Dune has grown to be a fairly large project that over time has acquired its own
+style. Below is an attempt to enumerate some important points of this style.
+These rules aren't axioms and we may break them when justified. However, we
+should have a good reason in mind when breaking them. Finally, the list isn't
+exhaustive by any means and is subject to change. Feel free to discuss anything
+in particular with the team.
+
+- Parameter signatures should be self descriptive. Use labels when the types
+  alone aren't sufficient to make the signature readable.
+
+Bad:
+
+.. code:: ocaml
+
+   val display_name : string -> string -> _ Pp.t
+
+Good:
+
+.. code:: ocaml
+
+   val display_name : first_name:string -> last_name:string -> _ Pp.t
+
+- Avoid type aliases when possible. Yes, they might make some type signatures
+  more readable, but they make the code harder to grep and make merlin's
+  inferred types more confusing.
+
+- Every .ml file must have a corresponding .mli. The only exception to this
+  rule is .ml files with only type definitions.
```

Comment:
Perhaps we should encourage only `mli` files with a type definition instead? I guess marking them in the stanza can be a bit more annoying however.

Reply:
I don't like those because one cannot define exceptions to them. I suppose should add that to the list